recipes/qt-advanced-docking-system/all/conanfile.py

```diff
@@ -26,70 +33,79 @@ class QtADS(ConanFile):
         "shared": False,
         "fPIC": True,
     }
-    generators = "cmake", "cmake_find_package", "cmake_find_package_multi"
-
-    _cmake = None
-    _qt_version = "5.15.6"
+    implements = ["auto_shared_fpic"]
 
     @property
-    def _source_subfolder(self):
-        return "source_subfolder"
-
-    def export_sources(self):
-        copy(self, "CMakeLists.txt", self.recipe_folder, self.export_sources_folder)
-        export_conandata_patches(self)
+    def _qt_major(self):
+        return Version(self.dependencies["qt"].ref.version).major
 
-    def config_options(self):
-        if self.settings.os == "Windows":
-            del self.options.fPIC
+    @property
+    def _min_cppstd(self):
+        if self._qt_major >= 6:
+            return 17
+        else:
+            return 14
 
-    def configure(self):
-        if self.options.shared:
-            del self.options.fPIC
+    def layout(self):
+        cmake_layout(self, src_folder="src")
 
     def requirements(self):
-        self.requires(f"qt/{self._qt_version}")
+        self.requires("qt/[>=6.0 <7]", transitive_headers=True, transitive_libs=True, run=can_run(self))
+        self.requires("libpng/[>=1.6 <2]")
 
-    def source(self):
-        get(self, **self.conan_data["sources"][self.version], strip_root=True,
-                  destination=self._source_subfolder)
+    def validate(self):
+        check_min_cppstd(self, self._min_cppstd)
 
-    def _configure_cmake(self):
-        if self._cmake:
-            return self._cmake
+    def build_requirements(self):
+        if not can_run(self):
+            self.tool_requires("qt/<host_version>", options={"gui": False, "widgets": False})
```

Comment:
```suggestion
    def build_requirements(self):
        if not can_run(self):
            self.tool_requires("qt/<host_version>", options={"gui": False, "widgets": False})
```

Why does Qt is needed as build requirement as well? It's true the project supports qmake as well, but I could not find any Qt macro in src/CMakeLists.txt. What am I missing?

Reply:
I'll double check later, but in the process of adding cross-compilation support to Qt I learned that it always looks for native Qt executables in its .cmake modules and fails if they are not found, even if they are not eventually used anywhere.
Currently, in recipes consuming Qt, the executables are found simply based on their location relative to the .cmake modules, even if they have not been added to PATH. That is why building against shared Qt fails in some recipes that are using the executables - the necessary shared libraries have not been added to env vars.
To fix that in the native build context, the Qt dep will need a `run=True` and/or `VirtualRunEnv(self).generate(scope="build")`. Always adding Qt as a tool_requires would also work, but I'm sure you can imagine how fun it is to build Qt twice.

A somewhat related question - is `self.requires(..., run=True)` property propagated in the graph as well or does it apply only to the local recipe. I think it might have been transitive. In that case generating VirtualBuildEnv for build would be more correct and safer.

Reply:
@valgur Thank you for your explanation. For this specific recipe there is no register about problems related to cross-building, so, could you please remove it?

Plus, Qt6 does not support cross-building for now. Neither libmysqlclient/8.1.0 (another transitive dependency), so for this package is not worth adding extra complexity regarding cross-building.

Reply:
Well... of course. Consider this a bug report for the cross-compilation issues. I can share a successful build log as well.

Reply:
@valgur That's not needed. Let's do that with the minimal PR that does that, that way we can move forward.  We can look at the cross-building situation in a follow-up PR, in particular, because if not can_run(self) is an unusual pattern to guard against for a tool requires.

Reply:
The `not can_run(self)` will effectively never run until Qt is cross-compilable, so there's no harm in it. Can you not trust me on this? I've been through the guts of the Qt recipe and its .cmake modules. The only alternatives are to heavily patch the modules to not check for the executables every time or to unconditionally always add Qt as tool_requires.

I understand that you won't like me wasting your time on this, but the feeling is mutual.

Reply:
fine